### PR TITLE
[Loot] Remove from shared memory, simplification

### PR DESCRIPTION
**Pull request by Akkadius**
Loot in shared memory is a cumbersome experience for server operators. 

This PR implements the following

* Remove loot from shared memory entirely
* Loads loot tables in bulk on initial zone spawn for relevant NPC's
* Lazy loads loot tables not loaded in initial bulk spawn when not found during dynamic quest spawns or otherwise
* Allows operators to reload loot via #reload loot or via Spire API globally without server restart. This does require a repop in order to re-roll loop if you want to see results
* Cleans up legacy loot code, simplifying and reorganizing code, utilizing repositories.

**Testing**

Loot works exactly as it did as expected.

image

image

**Reloading**

image

image

image

### CODE REVIEW

File: zone/common.h [hunk 1]
```diff
@@ -929,5 +929,6 @@ struct DataBucketCache
 	uint32_t      bucket_expires;
 };
 
+
```
Comment (Kinglykrab) [inline zone/common.h:932]: ?

File: zone/npc.h [hunk 2]
```diff
@@ -204,39 +208,37 @@ class NPC : public Mob
...
+	inline uint32 GetCopper() const { return m_loot_copper; }
+	inline uint32 GetSilver() const { return m_loot_silver; }
+	inline uint32 GetGold() const { return m_loot_gold; }
+	inline uint32 GetPlatinum() const { return m_loot_platinum; }
+	inline void SetCopper(uint32 amt) { m_loot_copper = amt; }
+	inline void SetSilver(uint32 amt) { m_loot_silver = amt; }
+	inline void SetGold(uint32 amt) { m_loot_gold = amt; }
+	inline void SetPlatinum(uint32 amt) { m_loot_platinum = amt; }
+
+	void DescribeAggro(Client *to_who, Mob *mob, bool verbose);
```
Comment (Kinglykrab) [inline zone/npc.h:240]: Where did `DescribeAggro()` come from?

Comment (Kinglykrab) [reply zone/npc.h:240]: Nevermind.

**Review by Kinglykrab — APPROVED** [review]
Looks great, glad to see this happen after all these years.

File: zone/zone_loot.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	if (loottable_ids.size() > 1) {
+		LogLoot("Loaded [{}] loottables", m_loottables.size());
```
Comment (hgtw) [inline zone/zone_loot.cpp:111]: Can we add a timer elapsed to this log

File: zone/pets.cpp [hunk 4]
```diff
@@ -576,7 +576,7 @@ void NPC::SetPetState(SpellBuff_Struct *pet_buffs, uint32 *items) {
 			bool petCanHaveNoDrop = (RuleB(Pets, CanTakeNoDrop) && _CLIENTPET(this) && GetPetType() <= petOther);
 
 			if (!noDrop || petCanHaveNoDrop) {
-				AddLootDrop(item2, &itemlist, NPC::NewLootDropEntry(), true);
+				AddLootDrop(item2, LootdropEntriesRepository::NewEntity(), true);
```
[the remark this answers is not in the document]
Comment (germboy007) [reply zone/pets.cpp:579]: Not sure if totally related, but testing out this code I've noticed pets don't always retain items when zoning. At first I thought it was certain zones, then it became all zones. Seems like it started with and got worse after using #reload loot

[the remark this answers is not in the document]
Comment (Akkadius) [reply zone/pets.cpp:579]: That might’ve been an issue prior to this though

[the remark this answers is not in the document]
Comment (Akkadius) [reply zone/pets.cpp:579]: > The old `NewLootDropEntry` had different default values than `NewEntity`. Specifically `chance` was 0, `multiplier` was 0, and `equip_item` was 1 in the old calls.
> 
> Would `equip_item` defaulting to 0 mean trading pets items won't make them equip the items now?

I’ll check out defaults, good eye

File: zone/zone_loot.cpp (hunk 3 above)
Comment (Akkadius) [reply zone/zone_loot.cpp:111]: Sure thing. It’s fairly lightweight with this mostly happening in zone boot

Comment (Akkadius) [reply zone/zone_loot.cpp:111]:
```
  Zone |    Info    | LoadLootTables Loaded [56] loottables (0.000302s)
```

File: zone/pets.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Akkadius) [reply zone/pets.cpp]: Swapped defaults for originals

**Review by hgtw — APPROVED** [review]
Couple memory leaks that existed in the old code and was carried over with the file rename, otherwise didn't notice anything

File: zone/loot.cpp [hunk 5]
```diff
@@ -0,0 +1,910 @@
...
+		augment_one,
+		augment_two,
+		augment_three,
+		augment_four,
+		augment_five,
+		augment_six
+	);
+
+	if (!inst) {
+		return;
```
[the remark this answers is not in the document]
Comment (hgtw) [reply zone/loot.cpp:337]: For a quick fix on these leaks, this one should probably just call `CreateItem` and then bail if it fails at the top of the function before `new LootItem` is created. The return in the middle of a big function like this is too error prone and it doesn't look like CreateItem needs anything except the function's arguments.

File: zone/loot.cpp [hunk 6]
```diff
@@ -0,0 +1,910 @@
...
+			}
+		}
+
+		if (found) {
+			item->equip_slot = found_slot;
+		}
+	}
+
+	if (found_slot != INVALID_INDEX) {
+		GetInv().PutItem(found_slot, *inst);
```
[the remark this answers is not in the document]
Comment (hgtw) [reply zone/loot.cpp:481]: To fix this leak a `safe_delete(inst)` at the end of the function should be enough

**Review by hgtw — COMMENTED** [review]
Added suggested fixes for the leaks

Unrelated to the PR but this function probably needs to be investigated though. It's weird that it adds item instances to the mob's inventory and then also adds to a separate list with a struct that looks like it holds most of the same data as an ItemInstance. Maybe there's some legacy reasons or inventory can't be used for drops?

**Pull request merged**